sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala

```diff
@@ -175,6 +175,27 @@ case class DataSourceResolution(
     case DataSourceV2Relation(CatalogTableAsV2(catalogTable), _, _) =>
       UnresolvedCatalogRelation(catalogTable)
 
+    case ShowTablesStatement(None, pattern) =>
+      defaultCatalog match {
+        case Some(_) =>
+          throw new AnalysisException(
+            "The default namespace cannot be deduced in v2 catalog yet")
+        case None =>
+          ShowTablesCommand(None, pattern)
+      }
+
+    case plan @ ShowTablesStatement(Some(namespace), pattern) =>
+      val CatalogNamespace(maybeCatalog, ns) = namespace
+      maybeCatalog match {
+        case Some(v2Catalog) =>
+          if (ns.isEmpty) {
+            throw new AnalysisException(
+              "The default namespace cannot be deduced in v2 catalog yet")
+          }
+          ShowTables(plan.output, v2Catalog.asTableCatalog, ns, pattern)
+        case None =>
+          ShowTablesCommand(Some(namespace.quoted), pattern)
```

Comment:
we can only fallback to the v1 show tables if `namespace.length == 1`. Otherwise we should fail and ask users to specify a valid database name.

`SHOW TABLES a.b` fails in Spark 2.x, and it should still fail in Spark 3.0 if users do not set the default catalog.

Reply:
Added the check. Thanks for pointing it out.